feat: add onchain address validation for network-specific addresses

Camillarhi opened the pull request:
Addresses need to be validated against the target network before use to ensure correctness
and prevent potential loss of funds.

This commit introduces a function to validate addresses based on the target network:
- Implements a function to validate addresses based on the network.
- Adds unit tests to verify the correctness of the validation logic.

This change enhances the robustness of address handling and mitigates risks associated with invalid or mismatched addresses.

Fixes https://github.com/lightningdevkit/ldk-node/issues/386

enigbe commented:
> Yeah, honestly, I'm a bit on the fence. `rust-bitcoin`'s enforced validation via types is a bit cumbersome, and generally something I don't want to force down our user's throats, especially in bindings targets where it may be even less idiomatic. Maybe just making sure to validate the addresses before using them is the way to go, we just have to make sure we're not missing any codepaths.

Thanks for clarifying this.

Camillarhi commented on src/error.rs:
```diff
@@ -193,6 +201,8 @@ impl fmt::Display for Error {
 				write!(f, "The given operation failed due to the LSP's required opening fee being too high.")
 			},
+			Self::InvalidAddressFormat => write!(f, "The given address is invalid."),
+			Self::InvalidNetworkAddress { expected } => write!(f, "The given address is invalid. Expected network: {:?}", expected),
```
Got it, I will make the changes

Camillarhi commented on src/payment/onchain.rs:
````diff
@@ -80,12 +82,33 @@ impl OnchainPayment {
+	/// ```
+	/// let address = "tb1q..."; // Testnet address
+	/// parse_and_validate_address(Network::Testnet, address)?;
+	pub fn parse_and_validate_address(&self, network: Network, address: &str) -> Result<Address, Error> {
````
Thank you for pointing this out! I’ve updated the `parse_and_validate_address` function to accept a `bitcoin::Address` instead of a `&str`, as the address is already parsed at the call site.

I also noticed that the `send_all_to_address` function should similarly use `parse_and_validate_address`. I’ll update it to include the validation step, ensuring both `send_to_address` and `send_all_to_address` consistently validate the address against the expected network.

Let me know if there’s anything else you’d like me to adjust!

Camillarhi replied:
I am not sure I can access the network currently being used from the wallet crate

Camillarhi commented on src/payment/onchain.rs:
```diff
@@ -120,6 +126,25 @@ impl OnchainPayment {
+	pub fn parse_and_validate_address(
+		&self, network: Network, address: &Address,
+	) -> Result<Address, Error> {
+		Address::<NetworkUnchecked>::from_str(address.to_string().as_str())
```
Alright , do I go ahead and expose this functionality for the binding targets?

Camillarhi commented:
> Seems CI is still failing, please let me know when this is ready for another round of review.

Hello @tnull , this is ready for another review, though the CI is still failing

Camillarhi commented on tests/integration_tests_rust.rs:
```diff
@@ -302,6 +305,9 @@ fn onchain_send_receive() {
 
 	let addr_a = node_a.onchain_payment().new_address().unwrap();
 	let addr_b = node_b.onchain_payment().new_address().unwrap();
+	let static_address = "tb1q0d40e5rta4fty63z64gztf8c3v20cvet6v2jdh";
```
Sure, I'll do that

Camillarhi commented on src/wallet/mod.rs:
```diff
@@ -327,10 +331,30 @@ where
 		self.get_balances(total_anchor_channels_reserve_sats).map(|(_, s)| s)
 	}
 
+	/// Validates a Bitcoin address is properly formatted and matches the expected network.
```
Thanks. I will drop them

Camillarhi commented:
> Please excuse the delay here!
> 
> LGTM!
> 
> Please squash the fixup commits into the feature commits to cleanup the commit history. Feel free to address the nit below while squashing.
> 
> Note it's usually prerable to keep a clean commit history at all times, and to clearly mark the fixup commits as you go. This link is a good starting point for some guidance on how to write commit messages: https://cbea.ms/git-commit/

Thanks, I will get on this and tag you as soon as it's done

Camillarhi commented:
> Please excuse the delay here!
> 
> LGTM!
> 
> Please squash the fixup commits into the feature commits to cleanup the commit history. Feel free to address the nit below while squashing.
> 
> Note it's usually prerable to keep a clean commit history at all times, and to clearly mark the fixup commits as you go. This link is a good starting point for some guidance on how to write commit messages: https://cbea.ms/git-commit/

Hello @tnull , I have squashed the commits and cleaned up the commit history.
